exoscale/api/compute.py

```diff
@@ -0,0 +1,2088 @@
+
+        Returns:
+            None
+        """
+
+        try:
+            if self.type == "ingress":
+                self.compute.cs.revokeSecurityGroupIngress(id=self.id)
+            else:
+                self.compute.cs.revokeSecurityGroupEgress(id=self.id)
+        except CloudStackApiException as e:
+            raise APIException(e.error["errortext"], e.error)
+
+        self.type = None
+        self.id = None
+        self.description = None
+        self.network_cidr = None
+        self.security_group = None
+        self.port = None
+        self.protocol = None
+        self.icmp_code = None
+        self.icmp_type = None
+
+    def _parse_port(self):
+        """
+        Parse the rule port attribute and returns a port range.
+
+        Returns:
+            int: start port
+            int: end port
+        """
+
+        if not self.port:
+            return None, None
+
+        start_port, end_port = (
+            self.port.split("-", maxsplit=1) if "-" in self.port else (self.port, None)
+        )
+        if not end_port:
+            end_port = start_port
+
+        return int(start_port), int(end_port)
+
+
+@attr.s
+class SSHKey(Resource):
+    """
+    A SSH key.
+
+    Attributes:
+        name (str): the SSH key unique name
+        fingerprint (str): the SSH key fingerprint
+        private_key (str): the SSH private key, or None if registered SSH key
+    """
+
+    compute = attr.ib(repr=False)
+    res = attr.ib(repr=False)
+    name = attr.ib()
+    fingerprint = attr.ib()
+    private_key = attr.ib(default=None, repr=False)
+
+    @classmethod
+    def from_cs(cls, compute, res):
+        return cls(
+            compute,
+            res,
+            name=res["name"],
+            fingerprint=res["fingerprint"],
+            private_key=res.get("privatekey", None),
+        )
+
+    def delete(self):
+        """
+        Delete the SSH key.
+
+        Returns:
+            None
+        """
+
+        try:
+            self.compute.cs.deleteSSHKeyPair(name=self.name)
+        except CloudStackApiException as e:
+            raise APIException(e.error["errortext"], e.error)
+
+        self.res = None
+        self.name = None
+        self.fingerprint = None
+        self.private_key = None
+
+
+@attr.s
+class Zone(Resource):
+    """
+    An Exoscale zone.
+
+    Attributes:
+        id (str): the zone unique identifier
+        name (str): the zone name
+    """
+
+    res = attr.ib(repr=False)
+    id = attr.ib()
+    name = attr.ib()
+
+    @classmethod
+    def from_cs(cls, res):
+        return cls(res, id=res["id"], name=res["name"])
+
+
+class ComputeAPI(API):
+    """
+    An Exoscale Compute API client.
+
+    Parameters:
+        key (str): the Compute API key
+        secret (str): the Compute API secret
+        endpoint (str): the Compute API endpoint
+        trace (bool): API request/response tracing flag
+    """
+
+    def __init__(self, key, secret, endpoint=None, trace=False):
```

Comment:
```suggestion
    def __init__(self, key, secret, endpoint="https://api.exoscale.com/compute", trace=False):
```

(a string isn't mutable, so it's fine as a default keyword argument)

Reply:
I'm probably doing something wrong, but when I do this the constructor `endpoint` parameter value is actually `None` and so I get the following error at runtime 🤔

```
requests.exceptions.MissingSchema: Invalid URL 'None': No schema supplied. Perhaps you meant http://None?
```

Reply:
Changed in d420474.